library/std/src/sys/configure_builtins.rs

```diff
@@ -1,8 +1,35 @@
-/// Hook into .init_array to enable LSE atomic operations at startup, if
-/// supported.
-#[cfg(all(target_arch = "aarch64", target_os = "linux", not(feature = "compiler-builtins-c")))]
+/// Enable LSE atomic operations at startup, if supported.
+///
+/// Linker sections are based on what [`ctor`] does, with priorities to run slightly before user
+/// code:
+///
+/// - Apple uses the section `__mod_init_func`, `mod_init_funcs` is needed to set
+///   `S_MOD_INIT_FUNC_POINTERS`. There doesn't seem to be a way to indicate priorities.
+/// - Windows uses `.CRT$XCT`, which is run before user constructors (these should use `.CRT$XCU`).
+/// - ELF uses `.init_array` with a priority of 90, which runs before our `ARGV_INIT_ARRAY`
+///   initializer (priority 99). Both are within the 0-100 implementation-reserved range, per docs
+///   for the [`prio-ctor-dtor`] warning, and this matches compiler-rt's `CONSTRUCTOR_PRIORITY`.
+///
+/// To save startup time, the initializer is only run if outline atomic routines from
+/// compiler-builtins may be used. If LSE is known to be available then the calls are never
+/// emitted, and if we build the C intrinsics then it has its own initializer using the symbol
+/// `__aarch64_have_lse_atomics`.
+///
+/// [`ctor`]: https://github.com/mmastrac/rust-ctor/blob/63382b833ddcbfb8b064f4e86bfa1ed4026ff356/shared/src/macros/mod.rs#L522-L534
+/// [`prio-ctor-dtor`]: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
```

Comment:
Reflecting on it, I think it'd be useful with a comment here why we do this in a static initialiser, and aren't doing it in `std`'s `init` function in `rt.rs`?

I can think of an answer myself: because this affects atomics, which are in `core`, and because it'd be confusing if `#![no_main]` negatively affected atomics performance - but in that case, why aren't we doing the initialisation in core to start with, to not affect `#![no_std]`? The answer here is probably because we need runtime feature detection, which is only available in `std`.

But again, it'd be nice to have a discussion about this in the docs.

Reply:
I think this would make sense as an addition.

Reply:
I didn't know of any strong reason for `init_array` other than following precedent; C compiler-rt does a global constructor as well, though doing it before main isn't really an option with how source is organized there. Our comment on `ARGV_INIT_ARRAY` has a good point though, that using `init_array` means things work even if we're in a dylib, or `no_main` (as you mentioned).

Feature detection is indeed the reason to be in std as aarch64 needs auxvec. `compiler-rt` [can do that](https://github.com/llvm/llvm-project/blob/e66e8aac6e4c0cc3874cabad736233440b0ffb1e/compiler-rt/lib/builtins/cpu_model/aarch64.c#L37-L54) but it would be a mess trying to depend on `std` or `libc` symbols from compiler-builtins.

I added a comment to this effect.